examples/hubert/dataset/hubert_dataset.py

```diff
@@ -0,0 +1,330 @@
+        batch_size (int or None, optional): The number of samples in one mini-batch.
+             (Default: ``None``)
+
+    Note: If ``max_token_count`` is not ``None``, the ``batch_size`` couldn't be set since
+        the lengths of samples are unknown, the batch size may be different for different
+        mini-batches.
+    """
+    def __init__(
+        self,
+        data_source: Dataset,
+        num_buckets: int,
+        max_token_count: Optional[int] = None,
+        batch_size: Optional[int] = None
+    ) -> None:
+        if max_token_count is not None and batch_size is not None:
+            raise AssertionError(
+                "The ``max_token_count`` and ``batch_size`` can't be both set."
+            )
+        self.data_source = data_source
+        self.max_token_count = max_token_count
+        self.buckets = self._get_buckets(self.data_source, num_buckets)
+
+    def _get_buckets(
+        self,
+        data_source: Dataset,
+        num_buckets: int
+    ) -> Dict[int, Tensor]:
+        """Generate buckets based on the dataset.
+        Args:
+            data_source (Dataset): The dataset object to bucketize.
+            num_buckets (int): The number of buckets.
+
+        Returns:
+            (dict[int, Tensor]): A dictionary in which the key is the bucket index, the value is
+                the Tensor of corresponding sample indices.
+        """
+        buckets = {}
+        len_list = data_source.len_list
+        min_len, max_len = min(len_list), max(len_list)
+
+        boundaries = [min_len - 1]
+        interval = (max_len - min_len) // num_buckets
+        for i in range(1, num_buckets):
+            boundaries.append(min_len + i * interval)
+        boundaries.append(max_len + 1)
+        bucket_ids = torch.bucketize(torch.Tensor(len_list), torch.Tensor(boundaries))
+        for i, _ in enumerate(len_list):
+            bucket_id = bucket_ids[i]
+            if bucket_id in buckets:
+                buckets[bucket_id].append(i)
+            else:
+                buckets[bucket_id] = [i]
+        for k in buckets:
+            random.shuffle(buckets[k])
+            buckets[k] = torch.as_tensor(buckets[k], dtype=torch.int)
+        return buckets
+
+    def __iter__(self) -> Iterator[List[int]]:
+        iter_list = []
+        total_len = 0
+        batch = []
+        len_list = self.data_source.len_list
+        if self.max_token_count:
+            for k in self.buckets.keys():
+                for i in range(self.buckets[k].size(0)):
+                    index = self.buckets[k][i]
+                    if total_len + len_list[index] > self.max_token_count:
+                        iter_list.append(batch)
+                        batch = [index]
+                        total_len = len_list[index]
+                    else:
+                        batch.append(index)
+                        total_len += len_list[index]
+        else:
+            for i in range(self.buckets[k].size(0)):
```

Comment:
Please correct me if I'm wrong. Where does `k` come from in this clause?

Reply:
Oops, I didn't test the batch size option. I will fix it.